internal/pkg/image/packer/gocryptfs.go

```diff
@@ -0,0 +1,207 @@
+package packer
+
+import (
+	"bytes"
+	"crypto/md5"
+	"crypto/rand"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"os/signal"
+	"path/filepath"
+	"strings"
+	"syscall"
+
+	"github.com/apptainer/apptainer/internal/pkg/image/driver"
+	"github.com/apptainer/apptainer/internal/pkg/util/bin"
+	"github.com/apptainer/apptainer/pkg/image"
+	"github.com/apptainer/apptainer/pkg/sylog"
+	"github.com/apptainer/apptainer/pkg/util/apptainerconf"
+	"github.com/apptainer/apptainer/pkg/util/cryptkey"
+)
+
+type Gocryptfs struct {
+	*Squashfs
+	gocryptfsPath string
+	driver        image.Driver
+	Pass          string
+	keyInfo       *cryptkey.KeyInfo
+}
+
+type cryptInfo struct {
+	cipherDir, plainDir, pass, confPath string
+}
+
+func newCryptInfo() *cryptInfo {
+	return &cryptInfo{
+		cipherDir: "",
+		plainDir:  "",
+		pass:      "",
+		confPath:  "",
+	}
+}
+
+func NewGocryptfs(keyInfo *cryptkey.KeyInfo) *Gocryptfs {
+	g := &Gocryptfs{
+		Squashfs: NewSquashfs(),
+	}
+	appfile := apptainerconf.GetCurrentConfig()
+	driver.InitImageDrivers(true, true, appfile, image.ImageFeature)
+	g.gocryptfsPath, _ = bin.FindBin("gocryptfs")
+	g.driver = image.GetDriver(driver.DriverName)
+	g.keyInfo = keyInfo
+	return g
+}
+
+func (g *Gocryptfs) HasGocryptfs() bool {
+	return g.driver.Features()&image.ImageFeature != 0 && g.gocryptfsPath != ""
+}
+
+func (g *Gocryptfs) init(tmpDir string) (cryptInfo *cryptInfo, err error) {
+	if !g.HasGocryptfs() {
+		return nil, fmt.Errorf("imagedriver is not initialized")
+	}
+
+	cryptInfo = newCryptInfo()
+	parentDir, err := os.MkdirTemp(tmpDir, "gocryptfs-")
+	if err != nil {
+		return
+	}
+	cipherDir := filepath.Join(parentDir, "cipher")
+	plainDir := filepath.Join(parentDir, "plain")
+
+	err = os.Mkdir(cipherDir, 0o700)
+	if err != nil {
+		return
+	}
+	cryptInfo.cipherDir = cipherDir
+	err = os.Mkdir(plainDir, 0o700)
+	if err != nil {
+		return
+	}
+	cryptInfo.plainDir = plainDir
+
+	buf := make([]byte, 32)
+	_, err = rand.Read(buf)
+	if err != nil {
+		return
+	}
+
+	switch g.keyInfo.Format {
+	case cryptkey.PEM:
+		hash := md5.Sum(buf)
+		cryptInfo.pass = hex.EncodeToString(hash[:])
+	case cryptkey.Passphrase:
+		cryptInfo.pass = g.keyInfo.Material
+	default:
+		err = errors.New("cryptkey type is unknown")
+		return
+	}
+	cryptInfo.confPath = filepath.Join(cipherDir, "gocryptfs.conf")
+
+	errCh := make(chan error, 1)
+	go func() {
```

Comment:
I'm very happy to see this using the image driver now.  I wonder if this go routine here is still needed, since the image driver Mount already forks a process?

Reply:
resolved in this commit
https://github.com/apptainer/apptainer/pull/1185/commits/17c45e4d5d0eab05651c9877307fcb7d58282232